cephfs/block_diff_test.go

```diff
@@ -0,0 +1,166 @@
+	}()
+
+	snap2 := "Snap2"
+	err = fsa.CreateSubVolumeSnapshot(volume, NoGroup, subname, snap2)
+	assert.NoError(t, err)
+	defer func() {
+		err := fsa.RemoveSubVolumeSnapshot(volume, NoGroup, subname, snap2)
+		assert.NoError(t, err)
+	}()
+
+	// Nothing has changed between snap1 and snap2.
+	changedBlocks := getChangedBlocks(t, mount, subVolRootPath, relPath, snap1, snap2)
+	assert.NotNil(t, changedBlocks)
+	assert.Equal(t, 0, len(*changedBlocks))
+
+	_, err = rand.Read(randData) // read 4MB of random data
+	assert.NoError(t, err)
+
+	// Write 4MB random data to the file.
+	f1, err = mount.Open(f1path, os.O_RDWR, 0644)
+	assert.NoError(t, err)
+	assert.NotNil(t, f1)
+	_, err = f1.WriteAt(randData, 0)
+	assert.NoError(t, err)
+	assert.NoError(t, f1.Sync())
+	assert.NoError(t, f1.Close())
+
+	snap3 := "Snap3"
+	err = fsa.CreateSubVolumeSnapshot(volume, NoGroup, subname, snap3)
+	assert.NoError(t, err)
+	defer func() {
+		err := fsa.RemoveSubVolumeSnapshot(volume, NoGroup, subname, snap3)
+		assert.NoError(t, err)
+	}()
+
+	// 4MB at offset 0 has changed between snap2 and snap3.
+	changedBlocks = getChangedBlocks(t, mount, subVolRootPath, relPath, snap2, snap3)
+	assert.NotNil(t, changedBlocks)
+	assert.Equal(t, len(*changedBlocks), 1)
+	assert.Equal(t, (*changedBlocks)[0].Offset, uint64(0))
+	assert.Equal(t, (*changedBlocks)[0].Len, uint64(4*1024*1024))
+
+	// read 4MB of random data
+	_, err = rand.Read(randData)
+	assert.NoError(t, err)
+
+	// Write at multiple intervals to the file.
+	f1, err = mount.Open(f1path, os.O_RDWR, 0644)
+	assert.NoError(t, err)
+	assert.NotNil(t, f1)
+	// write 32 1 byte random data at offset 0 with gap of 20 bytes
+	for i := 0; i < 32; i++ {
+		_, err = f1.WriteAt(randData[i:i+1], int64(i*21))
+		assert.NoError(t, err)
+	}
+	assert.NoError(t, f1.Sync())
+	assert.NoError(t, f1.Close())
+
+	snap4 := "Snap4"
+	err = fsa.CreateSubVolumeSnapshot(volume, NoGroup, subname, snap4)
+	assert.NoError(t, err)
+	defer func() {
+		err := fsa.RemoveSubVolumeSnapshot(volume, NoGroup, subname, snap4)
+		assert.NoError(t, err)
+	}()
+
+	// test for multiple changed blocks between snap3 and snap4.
+	changedBlocks = getChangedBlocks(t, mount, subVolRootPath, relPath, snap3, snap4)
+	assert.NotNil(t, changedBlocks)
+	assert.Equal(t, len(*changedBlocks), 32)
+}
+
+func getChangedBlocks(t *testing.T,
+	mount *MountInfo,
+	rootPath, relPath, snap1, snap2 string) *[]CBlock {
+	fileBlockDiffInfo, err := FileBlockDiffInit(mount, rootPath, relPath, snap1, snap2)
+	assert.NoError(t, err)
+	changedBlocksList := make([]CBlock, 0)
+	defer func() {
+		assert.NoError(t, fileBlockDiffInfo.Close())
+	}()
+	for {
+		noMoreEntries, fileChangedBlocksInfo, err := fileBlockDiffInfo.ReadFileBlockDiff()
+		assert.NoError(t, err)
+		if fileChangedBlocksInfo == nil || fileChangedBlocksInfo.NumBlocks == 0 {
+			break
+		}
+		changedBlocksList = append(changedBlocksList, fileChangedBlocksInfo.CBlocks...)
+		if noMoreEntries {
+			break
+		}
```

Comment:
Do we ever exit the loop with this `break` statement? IOW, does a value 0 for _fileChangedBlocksInfo.NumBlocks_ indicate that there are no more entries left to read using `ReadFileBlockDiff()`?